### PR TITLE
Fix numerical instability of tf.norm gradient (NaN at zero)

### LINKED ISSUE #12071: Numerical instability of gradient calculation of tf.norm (nan at 0, inf for small values)
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes see below
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Mac OS X 10.11.6
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: #v1.2.0-5-g435cdfc    1.2.1
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:
- **CUDA/cuDNN version**: On CPU
- **GPU model and memory**:
- **Exact command to reproduce**: tf.norm at [0,0] see below for code

```
import numpy as np
import tensorflow as tf
print(tf.GIT_VERSION, "  ", tf.VERSION) #v1.2.0-5-g435cdfc    1.2.1

X = tf.placeholder(tf.float32, shape=(4,None))
Z = tf.norm(X, ord='euclidean', axis=1, name='logit')
var_grad = tf.gradients(Z, [X])

with tf.Session() as sess:
    X_ = np.array([
        [1],  # Grad OK
        [0],  # Grad NaN
        [1e-16],  # Grad OK
        [1e-19] #Grad Inf
    ], dtype=np.float32)
    sess.run(tf.global_variables_initializer())
    print(sess.run((Z, var_grad), feed_dict={X: X_}))
    # Result:
    #(array([9.99999940e-01, 0.00000000e+00, 9.99999951e-17,
    #        0.00000000e+00], dtype=float32), [array([[1.00000012],
    #                                                 [nan],
    #                                                 [1.],
    #                                                 [inf]], dtype=float32)])
```

### Describe the problem
`nan` is calculated for the gradient of `tf.norm` at zero values. For extremely small values `inf` is calculated. Note that the exact result should be 1 in all cases above. 

Above is a minimal example to reproduce it. The problem occurred in a real world scenario, when implementing a custom loss function (the entropy in https://arxiv.org/abs/1611.01449) and two embeddings where too close to each other (distance practically 0).

### Source code / logs
See above 

#### Output of logfile
```
== cat /etc/issue ===============================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64
Mac OS X 10.11.6

== are we in docker =========================================  echo == are we in docker ====================================num echo == are we in docker =========================================  ec==  echo == are we in docker =======================================c++ --version

== uname -a =====================================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy (1.13.0)
protobuf (3.3.0)
tensorflow (1.2.1)

== check for virtualenv ==============  echo == check for virtualenv =====on_b echo == check fo sys  echo == check for virtualenv ============== echo == check for virtualenv ============================================

== cat /etc/issue ===============================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64
Mac OS X 10.11.6

== are we in docker =============================================
No

== compiler =====================================================
Apple LLVM version 7.3.0 (clang-703.0.31)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

== uname -a =====================================================
Darwin Olivers-MBP-5.fritz.box 15.6.0 Darwin Kernel Version 15.6.0: Tue Apr 11 16:00:51 PDT 2017; root:xnu-3248.60.11.5.3~1/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy (1.13.0)
protobuf (3.3.0)
tensorflow (1.2.1)

== check for virtualenv =========================================
True

== tensorflow import ============================================
tf.VERSION = 1.2.1
tf.GIT_VERSION = v1.2.0-5-g435cdfc
tf.COMPILER_VERSION = v1.2.0-5-g435cdfc
Sanity check: array([1], dtype=int32)

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
tf_env_collect.sh.txt: line 105: nvidia-smi: command not found

== cuda libs  ===================================================
```

### PR DESCRIPTION
**Pull request by ayushozha**
## Summary

- The gradient of `tf.norm` / `tf.math.reduce_euclidean_norm` produces **NaN** when the input is zero and **inf** for very small values
- The gradient formula `x / norm(x)` is `0/0 = NaN` at zero; the correct subgradient is **0** (matching JAX and PyTorch)
- This has been an open issue for 8+ years

## Changes

Two code paths are fixed:

**1. `_EuclideanNormGrad`** (`math_grad.py`) — used by `tf.math.reduce_euclidean_norm`:
- Replace `truediv(x, output / grad)` with `div_no_nan(x, output) * grad`
- `div_no_nan` returns 0 when the denominator (norm) is 0

**2. `tf.linalg.norm` / `tf.norm`** (`linalg_ops.py`) — uses `sqrt(reduce_sum(x*x))`:
- The `sqrt` gradient `0.5/sqrt(x)` is inf when `x=0`
- Use `maximum(sum_squares, tiny)` to keep the sqrt gradient finite
- `where` mask preserves exact zero forward output
- Only XLA-compatible ops used (no new op types)

**3. Tests** (`math_grad_test.py`):
- Updated `testZeros` to expect gradient of `[0.0, 0.0]` instead of `[NaN, NaN]`
- Added `LinalgNormGradientTest` with zero and non-zero gradient tests

## Reproducer

```python
import tensorflow as tf

# Path A: reduce_euclidean_norm
x = tf.constant([0.0, 0.0])
with tf.GradientTape() as tape:
    tape.watch(x)
    y = tf.math.reduce_euclidean_norm(x)
print(tape.gradient(y, x))  # Was: [nan, nan], Now: [0, 0]

# Path B: tf.norm
with tf.GradientTape() as tape:
    tape.watch(x)
    y = tf.norm(x)
print(tape.gradient(y, x))  # Was: [nan, nan], Now: [0, 0]
```

## Test plan

- [ ] Run `bazel test //tensorflow/python/ops:math_grad_test --test_filter=*EuclideanNorm*`
- [ ] Run `bazel test //tensorflow/python/ops:math_grad_test --test_filter=*LinalgNorm*`
- [ ] Verify non-zero inputs still produce correct gradients (covered by existing tests)

Fixes #12071

### CODE REVIEW

**Comment by ayushozha** [comment]
@keerthanakadiri I pushed an update for the Gemini feedback on this PR.

What changed:
- Made the safe `tf.linalg.norm` sqrt path dtype-aware for complex tensors by clamping the real sum and rebuilding a complex value for `sqrt`.
- Avoided `np.finfo` on `bfloat16` by using the float32 finfo for that real dtype.
- Extended the existing zero and non-zero gradient tests to cover `complex64` and `complex128`.

Validation run locally:
- `python -m py_compile tensorflow\python\ops\linalg_ops.py tensorflow\python\ops\math_grad.py tensorflow\python\ops\math_grad_test.py`
- `git diff --check`
- secret-pattern scan on the touched files

I could not run the Bazel target locally because `bazel` is not installed in this environment.

File: tensorflow/python/ops/math_grad_test.py [hunk 1]
```diff
@@ -353,6 +353,33 @@ def test3D_4(self):
...
+    for dtype in [dtypes.float32, dtypes.float64]:
+      x = constant_op.constant([3.0, 4.0], dtype=dtype)
+
+      with backprop.GradientTape() as tape:
+        tape.watch(x)
+        y = linalg_ops.norm_v2(x)
+
+      dx = tape.gradient(y, x)
+      # Expected: x / norm(x) = [3/5, 4/5]
+      self.assertAllClose(dx, [0.6, 0.8])
```
[the remark this answers is not in the document]
Comment (ayushozha) [reply tensorflow/python/ops/math_grad_test.py:384]: Thanks @gemini-code-assist, addressed in `ad82a32f758`: `LinalgNormGradientTest.testZeroGrad` and `testNonZeroGrad` now cover `dtypes.complex64` and `dtypes.complex128` in addition to the real dtypes.

Validation: `python -m py_compile tensorflow/python/ops/math_grad_test.py`; `git diff --check`; PR `check-changes` and `cla/google` are passing.

**Comment by keerthanakadiri** [comment]
Hi @ayushozha , Can you kindly resolve the conflicts? Thank you!

**Comment by ratgr** [comment]
Hi @ayushozha, can you take a look at the conflict and resolve them? Thanks!

**Comment by keerthanakadiri** [comment]
Hi @ayushozha , Can you kindly resolve the conflicts? Thank you!
